src/librustdoc/html/render/mod.rs

```diff
@@ -1945,21 +1946,35 @@ pub(crate) fn small_url_encode(s: String) -> String {
     }
 }
 
-fn get_id_for_impl(for_: &clean::Type, trait_: Option<&clean::Path>, cx: &Context<'_>) -> String {
-    match trait_ {
-        Some(t) => small_url_encode(format!("impl-{:#}-for-{:#}", t.print(cx), for_.print(cx))),
-        None => small_url_encode(format!("impl-{:#}", for_.print(cx))),
-    }
+fn get_id_for_impl<'tcx>(tcx: TyCtxt<'tcx>, impl_id: ItemId) -> String {
+    let (type_, trait_) = match impl_id {
+        ItemId::Auto { trait_, for_ } => {
+            let ty = tcx.type_of(for_).skip_binder();
+            (ty, Some(tcx.mk_trait_ref(trait_, [ty])))
+        }
+        ItemId::Blanket { impl_id, .. } => (
+            tcx.type_of(impl_id).skip_binder(),
+            tcx.impl_trait_ref(impl_id).map(|x| x.skip_binder()),
+        ),
+        ItemId::DefId(impl_id) => (
+            tcx.type_of(impl_id).skip_binder(),
+            tcx.impl_trait_ref(impl_id).map(|x| x.skip_binder()),
+        ),
+    };
+    use rustc_middle::ty::print::with_forced_trimmed_paths;
```

Comment:
If you don't want to put it at the top of the file, can you at least put it at the beginning of the function body please?